optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,309 @@
+MAX_INT_EXHAUSTIVE_SEARCH_PARAMS = 16
+
+
+def _local_search_discrete(
+    acqf_params: AcquisitionFunctionParams,
+    initial_params: np.ndarray,
+    initial_fval: float,
+    param_idx: int,
+    choices: np.ndarray,
+    xtol: float,
+) -> tuple[np.ndarray, float] | None:
+    scale_type = acqf_params.search_space.scale_types[param_idx]
+    if scale_type == ScaleType.CATEGORICAL or len(choices) <= MAX_INT_EXHAUSTIVE_SEARCH_PARAMS:
+        return _local_search_discrete_exhaustive_search(
+            acqf_params, initial_params, initial_fval, param_idx, choices
+        )
+    else:
+        return _local_search_discrete_line_search(
+            acqf_params, initial_params, initial_fval, param_idx, choices, xtol
+        )
+
+
+def local_search_mixed(
+    acqf_params: AcquisitionFunctionParams,
+    initial_normalized_params: np.ndarray,
+    *,
+    tol: float = 1e-4,
+    max_iter: int = 100,
+) -> tuple[np.ndarray, float]:
+    scale_types = acqf_params.search_space.scale_types
+    bounds = acqf_params.search_space.bounds
+    steps = acqf_params.search_space.steps
+
+    continuous_params = np.where(steps == 0.0)[0]
+
+    inverse_squared_lengthscales = (
+        acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
+    )
+    # This is a technique for speeding up optimization.
+    # We use an isotropic kernel, so scaling the gradient will make
+    # the hessian better-conditioned.
+    continuous_param_scale = 1 / np.sqrt(inverse_squared_lengthscales[continuous_params])
+
+    noncontinuous_params = np.where(steps > 0)[0]
+    noncontinuous_param_choices = [
+        (
+            np.arange(bounds[i, 1])
+            if scale_types[i] == ScaleType.CATEGORICAL
+            else normalize_one_param(
+                param_value=np.arange(bounds[i, 0], bounds[i, 1] + 0.5 * steps[i], steps[i]),
+                scale_type=ScaleType(scale_types[i]),
+                bounds=(bounds[i, 0], bounds[i, 1]),
+                step=steps[i],
+            )
+        )
+        for i in noncontinuous_params
+    ]
+
+    noncontinuous_paramwise_xtol = [
+        # Enough to discriminate between two choices.
+        np.min(np.diff(choices), initial=np.inf) / 4
+        for choices in noncontinuous_param_choices
+    ]
+
+    normalized_params = initial_normalized_params.copy()
+    fval = float(eval_acqf_no_grad(acqf_params, normalized_params))
+
+    NEVER = -1
+    CONTINUOUS = -2
+    last_changed_param = NEVER
+
+    for _ in range(max_iter):
+        if last_changed_param == CONTINUOUS:
+            # Parameters not changed since last time.
+            return (normalized_params, fval)
+        res = _local_search_continuous(
+            acqf_params,
+            normalized_params,
+            fval,
+            continuous_params,
+            continuous_param_scale,
+            tol,
+        )
+        if res is not None:
+            normalized_params, fval = res
+            last_changed_param = CONTINUOUS
```

Comment:
If we update the return type, it becomes:
```suggestion
        last_changed_param = CONTINUOUS if updated else last_changed_param
```

I think it looks clearer as well.